gap/attributes/attract.gi

```diff
@@ -256,18 +256,32 @@ function(S)
 
   r := GreensRClassOfElementNC(S, rep);
 
-  if NrIdempotents(r) <> 1
-      or NrIdempotents(GreensDClassOfElementNC(S, rep)) <> 1 then
-    Info(InfoSemigroups, 2, "the number of idempotents in the D-class of the",
-         " first maximum rank");
-    Info(InfoSemigroups, 2, " generator is not 1");
+  if NrIdempotents(r) <> 1 or NrHClasses(r) <> 1 or
+      NrHClasses(GreensLClassOfElementNC(S, rep)) <> 1 then
+    Info(InfoSemigroups, 2, "the D-class of the first maximum rank generator ",
+                            "is not a group");
     return fail;
   fi;
 
   e := Idempotents(r)[1];
 
-  if ForAll(GeneratorsOfSemigroup(S), x -> x * e = x and e * x = x) then
-    return e;
+  if HasGeneratorsOfSemigroup(S) then
+    if ForAll(GeneratorsOfSemigroup(S), x -> x * e = x and e * x = x) then
+      return e;
+    fi;
+  elif ForAll(gens, x -> x * e = x and e * x = x) then
+    # S is an ideal without GeneratorsOfSemigroup
+    o := LambdaOrb(S);
+    Enumerate(o);
+    act := LambdaAct(S);
+    if ForAll([2 .. Length(o)], i -> act(o[i], e) = o[i]) then
```

Comment:
Same, why not negate and return false,

Reply:
Because you've put in `InfoStatement`s to the method that won't be triggered if I return early

Reply:
@james-d-mitchell Would you prefer me to remove the InfoStatement at the end of the method
```
  Info(InfoSemigroups, 2, "the unique idempotent in the D-class of the first",
                          " maximum rank");
  Info(InfoSemigroups, 2, " generator is not the identity");
  return fail;
```
and then restructure the code, with more `return fail`s, to reduce the nesting level?

Reply:
Yes probably, if you can be bothered!